### PR TITLE
skip checkpointing if there is no new data processed

### LINKED ISSUE #648: Skip checkpointing if there is no new data processed
**Describe what enhancement you'd like to have**
If there is no new data processed, we can skip new checkpoint request.

### PR DESCRIPTION
**Pull request by yl-lisen**
PR checklist:
- Did you run ClangFormat ?
- Did you separate headers to a different section in existing community code base ?
- Did you surround `proton: starts/ends` for new code in existing community code base ?

Please write user-readable short description of the changes:
This close #648

### CODE REVIEW

**Review by qijun-niu-timeplus — APPROVED** [review]

File: src/Checkpoint/CheckpointCoordinator.cpp [hunk 1]
```diff
@@ -387,58 +385,51 @@ void CheckpointCoordinator::removeExpiredCheckpoints(bool delete_marked)
...
-                        logger,
-                        "Failed to trigger checkpointing state for query={} epoch={}, since it's already cancelled",
-                        qid,
-                        next_epoch);
-                    return false;
-                }
-
-                exec->triggerCheckpoint(std::move(ckpt_ctx));
-            }
+    const auto & qid = ckpt_ctx->qid;
```
Comment (chenziliang) [inline src/Checkpoint/CheckpointCoordinator.cpp]: don't need to create this alias

File: src/Processors/Executors/ExecutingGraph.cpp [hunk 2]
```diff
@@ -543,6 +544,17 @@ void ExecutingGraph::initCheckpointNodes()
     assert(!checkpoint_trigger_nodes.empty() && !checkpoint_ack_nodes.empty());
 }
 
+bool ExecutingGraph::hasProcessedToCheckpoint() const
```
Comment (chenziliang) [inline src/Processors/Executors/ExecutingGraph.cpp]: Let's rename it as `hasProcessedNewDataSinceLastCheckpoint()`

**Comment by chenziliang** [comment]
CI failed

**Comment by yokofly** [comment]
I locally build release type with this branch and run smoke encounter a crash.
```
python3 ci_runner.py --local --no_retry -q random_stream 
```

```cpp
2024.04.09 19:05:24.286365 [ 19362 ] {2230} <Information> PipelineExecutor: Using 1 threads to execute pipeline for query_id=2230
2024.04.09 19:05:24.286692 [ 19362 ] {2230} <Information> LocalFileSystemCheckpoint: Took 0 ms to checkpoint to ./proton-data/checkpoint/2230/dag.ckpt, compressed_size=900, uncompressed_size=718
2024.04.09 19:05:24.286734 [ 19362 ] {2230} <Information> LocalFileSystemCheckpoint: Took 0 ms to checkpoint to ./proton-data/checkpoint/2230/query.ckpt, compressed_size=135, uncompressed_size=108
2024.04.09 19:05:24.286744 [ 19362 ] {2230} <Information> CheckpointCoordinator: Register query=2230 with 1 seconds checkpoint interval, source_node_descriptions=0-StreamingStoreSource, ack_node_descriptions=3-LazyOutputFormat
2024.04.09 19:05:24.299568 [ 19358 ] {c3efb4e2-1813-47c5-891c-0fa27ba73045} <Information> PipelineExecutor: Using 1 threads to execute pipeline for query_id=c3efb4e2-1813-47c5-891c-0fa27ba73045
2024.04.09 19:05:24.300126 [ 19066 ] {c3efb4e2-1813-47c5-891c-0fa27ba73045} <Information> executeQuery: Read 2 rows, 2.84 KiB in 0.00207991 sec., 961 rows/sec., 1.33 MiB/sec.
2024.04.09 19:05:25.287125 [ 19301 ] {} <Information> CheckpointCoordinator: Skipped checkpointing state for query=2230 epoch=1, since there is no new data processed
2024.04.09 19:05:26.287424 [ 19301 ] {} <Information> CheckpointCoordinator: Skipped checkpointing state for query=2230 epoch=1, since there is no new data processed
2024.04.09 19:05:27.288233 [ 24525 ] {} <Fatal> BaseDaemon: ########## Short fault info ############
2024.04.09 19:05:27.288297 [ 24525 ] {} <Fatal> BaseDaemon: (version 1.5.4, build id: A086D5E1E672C1F9700ABAAE6F36F224DAE7B656, git hash: 58465336fcf642f18b0277ea050dca9935b3182a) (from thread 19301) Received signal 11
2024.04.09 19:05:27.288318 [ 24525 ] {} <Fatal> BaseDaemon: Signal description: Segmentation fault
2024.04.09 19:05:27.288331 [ 24525 ] {} <Fatal> BaseDaemon: Address: 0x1f Access: read. Address not mapped to object.
2024.04.09 19:05:27.288348 [ 24525 ] {} <Fatal> BaseDaemon: Stack trace: 0x0000000019c6f8f9 0x0000000019c6f016 0x0000000019f627a7 0x0000000019f60c6d 0x0000000019f6098a 0x0000000019f5ea11 0x0000000019c76132 0x0000000010be83fd 0x0000000010beadd5 0x0000000010be662f 0x0000000010be906e 0x00007f89bea77609 0x00007f89be99c353
2024.04.09 19:05:27.288362 [ 24525 ] {} <Fatal> BaseDaemon: ########################################
2024.04.09 19:05:27.288371 [ 24525 ] {} <Fatal> BaseDaemon: (version 1.5.4, build id: A086D5E1E672C1F9700ABAAE6F36F224DAE7B656, git hash: 58465336fcf642f18b0277ea050dca9935b3182a) (from thread 19301) (no query) Received signal Segmentation fault (11)
2024.04.09 19:05:27.288378 [ 24525 ] {} <Fatal> BaseDaemon: Address: 0x1f Access: read. Address not mapped to object.
2024.04.09 19:05:27.288385 [ 24525 ] {} <Fatal> BaseDaemon: Stack trace: 0x0000000019c6f8f9 0x0000000019c6f016 0x0000000019f627a7 0x0000000019f60c6d 0x0000000019f6098a 0x0000000019f5ea11 0x0000000019c76132 0x0000000010be83fd 0x0000000010beadd5 0x0000000010be662f 0x0000000010be906e 0x00007f89bea77609 0x00007f89be99c353
2024.04.09 19:05:27.288444 [ 24525 ] {} <Fatal> BaseDaemon: 3. DB::CheckpointCoordinator::doTriggerCheckpoint(std::__1::weak_ptr<DB::PipelineExecutor> const&, std::__1::shared_ptr<DB::CheckpointContext const>) @ 0x0000000019c6f8f9 in ../build_rel/programs/proton
2024.04.09 19:05:27.288468 [ 24525 ] {} <Fatal> BaseDaemon: 4. DB::CheckpointCoordinator::triggerCheckpoint(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, unsigned long) @ 0x0000000019c6f016 in ../build_rel/programs/proton
2024.04.09 19:05:27.288485 [ 24525 ] {} <Fatal> BaseDaemon: 5. muduo::net::TimerQueue::handleRead() @ 0x0000000019f627a7 in ../build_rel/programs/proton
2024.04.09 19:05:27.288497 [ 24525 ] {} <Fatal> BaseDaemon: 6. muduo::net::Channel::handleEventWithGuard(muduo::Timestamp) @ 0x0000000019f60c6d in ../build_rel/programs/proton
2024.04.09 19:05:27.288508 [ 24525 ] {} <Fatal> BaseDaemon: 7. muduo::net::Channel::handleEvent(muduo::Timestamp) @ 0x0000000019f6098a in ../build_rel/programs/proton
2024.04.09 19:05:27.288523 [ 24525 ] {} <Fatal> BaseDaemon: 8. muduo::net::EventLoop::loop() @ 0x0000000019f5ea11 in ../build_rel/programs/proton
2024.04.09 19:05:27.288534 [ 24525 ] {} <Fatal> BaseDaemon: 9. DB::TimerService::startEventLoop() @ 0x0000000019c76132 in ../build_rel/programs/proton
2024.04.09 19:05:27.288557 [ 24525 ] {} <Fatal> BaseDaemon: 10. ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) @ 0x0000000010be83fd in ../build_rel/programs/proton
2024.04.09 19:05:27.288606 [ 24525 ] {} <Fatal> BaseDaemon: 11. decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&>(void&&) @ 0x0000000010beadd5 in ../build_rel/programs/proton
2024.04.09 19:05:27.288632 [ 24525 ] {} <Fatal> BaseDaemon: 12. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x0000000010be662f in ../build_rel/programs/proton
2024.04.09 19:05:27.288659 [ 24525 ] {} <Fatal> BaseDaemon: 13. void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x0000000010be906e in ../build_rel/programs/proton
2024.04.09 19:05:27.288685 [ 24525 ] {} <Fatal> BaseDaemon: 14. ? @ 0x00007f89bea77609 in ?
2024.04.09 19:05:27.288700 [ 24525 ] {} <Fatal> BaseDaemon: 15. __clone @ 0x00007f89be99c353 in ?
2024.04.09 19:05:27.288712 [ 24525 ] {} <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read.
2024.04.09 19:05:27.288727 [ 24525 ] {} <Information> SentryWriter: Not sending crash report
2024.04.09 19:05:27.309356 [ 19362 ] {2230} <Information> LocalFileSystemCheckpoint: Took 1 ms to checkpoint to ./proton-data/checkpoint/2230/1/0.ckpt, compressed_size=92, uncompressed_size=65
2024.04.09 19:05:27.309436 [ 19362 ] {2230} <Information> LocalFileSystemCheckpoint: Took 0 ms to checkpoint to ./proton-data/checkpoint/2230/1/1.ckpt, compressed_size=62, uncompressed_size=38
Segmentation fault (core dumped)

```

**Comment by yokofly** [comment]
both redp and nlog:
```
2024-04-10T06:02:02.1075946Z FAILED test_stream_smoke/test_query_smoke.py::test_run[two_level_global_aggr-3-subscribe-and-recover-two-level-global-aggr]
2024-04-10T06:02:02.1077589Z FAILED test_stream_smoke/test_query_smoke.py::test_run[changelog_stream13-103-global aggr on changelog_kv stream with single shard]
2024-04-10T06:02:02.1079219Z FAILED test_stream_smoke/test_query_smoke.py::test_run[changelog_stream13-107-global aggr on changelog_kv stream with single shard]
2024-04-10T06:02:02.1080734Z FAILED test_stream_smoke/test_query_smoke.py::test_run[query_state5-102-multiple_recover_global_aggr_in_substream]
2024-04-10T06:02:02.1081841Z ================== 4 failed, 839 passed in 1027.38s (0:17:07) ==================
```

```
2024-04-10T05:59:08.0823605Z =========================== short test summary info ============================
2024-04-10T05:59:08.0824853Z FAILED test_stream_smoke/test_query_smoke.py::test_run[two_level_global_aggr-3-subscribe-and-recover-two-level-global-aggr]
2024-04-10T05:59:08.0826472Z FAILED test_stream_smoke/test_query_smoke.py::test_run[changelog_stream13-103-global aggr on changelog_kv stream with single shard]
2024-04-10T05:59:08.0828115Z FAILED test_stream_smoke/test_query_smoke.py::test_run[changelog_stream13-107-global aggr on changelog_kv stream with single shard]
2024-04-10T05:59:08.0829619Z FAILED test_stream_smoke/test_query_smoke.py::test_run[query_state5-102-multiple_recover_global_aggr_in_substream]
2024-04-10T05:59:08.0830643Z ================== 4 failed, 846 passed in 858.35s (0:14:18) ===================
```

File: tests/stream/test_stream_smoke/0018_query_state5_substream.json [hunk 3]
```diff
@@ -93,17 +93,17 @@
                     "statements": [
                         {"client":"python", "query_type": "table", "query":"drop stream if exists test19_state_stream5"},
                         {"client":"python", "query_type": "table", "exist":"test19_state_stream5", "exist_wait":2, "wait":1, "query":"create stream test19_state_stream5 (id string, location string, value float, timestamp datetime64(3) default now64(3))"},
-                        {"client":"python", "query_type": "stream", "query_id":"19102", "wait":1, "terminate":"manual", "query":"subscribe to select id, count(*) over (partition by id), min(value) over (partition by id), max(value) over (partition by id) from test19_state_stream5 settings checkpoint_interval=1"},
+                        {"client":"python", "query_type": "stream", "query_id":"19102", "wait":1, "terminate":"manual", "query":"subscribe to select id, count(*) over (partition by id), min(value) over (partition by id), max(value) over (partition by id) from test19_state_stream5 settings checkpoint_interval=6"},
```
Comment (chenziliang) [inline tests/stream/test_stream_smoke/0018_query_state5_substream.json:96]: I think we will need fix the crash for frequent ckpt case or at least we will need open a bug for this if frequent ckpt cause crash and fix it in a separate PR.

Comment (yokofly) [reply tests/stream/test_stream_smoke/0018_query_state5_substream.json:96]: for now, the ci only fail in 4 tests( verify on my local, the value incorrect).
previous I indeed encounter a coredump.

File: src/Processors/Transforms/Streaming/AggregatingTransform.cpp [hunk 4]
```diff
@@ -549,6 +549,10 @@ void AggregatingTransform::recover(CheckpointContextPtr ckpt_ctx)
             {
                 UInt64 last_rows = 0;
                 DB::readIntBinary<UInt64>(last_rows, rb);
+                /// In case when for we had global aggregated some data, but done checkpoint request before finializing
```
Comment (chenziliang) [inline src/Processors/Transforms/Streaming/AggregatingTransform.cpp:552]: I feel like we don't need this logging and we will need fix the crash which is the significant part. With this logging, we can do nothing ?

**Pull request merged**